### PR TITLE
drjt Chapter 15 Exercises Complete

### PR DESCRIPTION
**Pull request by thruddle**

### CODE REVIEW

File: drjt/Animal/lib/Animal.pm [hunk 1]
```diff
@@ -38,6 +38,77 @@ if you don't export anything, such as for a purely object-oriented module.
...
+Constructor, takes a name as a parameter
+
+=cut
+
+sub new {
+  my $class = shift;
+  my $name = shift;
+  my $self = { name => $name, colour => $class->default_colour() };
+  bless $self, $class;
+  return $self;
```
Comment (SimonGross) [inline drjt/Animal/lib/Animal.pm:52]: You don't strictly need to have the object hash defined separately, unless you want to conduct additional processing after blessing it. You could simplify this to:

``` perl
return bless { name => $name, colour => $class->default_colour() }, $class;
```

File: drjt/Animal/lib/Animal.pm [hunk 2]
```diff
@@ -38,6 +38,77 @@ if you don't export anything, such as for a purely object-oriented module.
...
+A setter
+
+=cut
+
+sub set_name {
+  my $self = shift;
+  my $name = shift;
+  if ( ! ref $self ) {
+    die "Class has not been instantiated, Call a constructor first!\n";
+  }
```
Comment (SimonGross) [inline drjt/Animal/lib/Animal.pm:74]: This is a place where `croak`, which was demonstrated in the book, is more useful (from the perspective of this being an independent module). If we use `die` then we expose our internal workings a little, whereas `croak` causes the error to appear to have been thrown from the line that calls this method.

Comment (thruddle) [inline drjt/Animal/lib/Animal.pm:74]: Huh, must have missed that bit. But yeah, that's what the new-line is for, it prevents the `die` from including the module and line number - so not as useful, but it doesn't expose any workings.

File: drjt/Animal/lib/Animal.pm [hunk 3]
```diff
@@ -38,6 +38,77 @@ if you don't export anything, such as for a purely object-oriented module.
...
+  return $self;
+}
+
+=head2 get_name 
+
+=cut
+
+sub get_name {
+  my $self = shift;
+  ref $self ? $self->{name} : "Nameless $self";
```
Comment (Spinodal) [inline drjt/Animal/lib/Animal.pm:101]: not sure i like implicit returns and a ternary...

Comment (thruddle) [inline drjt/Animal/lib/Animal.pm:101]: Yeah, would probably be better if I made the `return` explicit.

**Pull request merged**